### PR TITLE
Clean up error object examples

### PR DESCRIPTION
**Pull request by ethanresnick**
Ok, so these changes got a bit more involved than I expected. I've left inline notes trying to explain each one. Lmk what you think!

re: https://github.com/json-api/json-api/pull/828

### CODE REVIEW

File: examples/index.md [hunk 1]
```diff
@@ -167,9 +167,11 @@ Content-Type: application/vnd.api+json
 
 Examples of how [error objects](http://jsonapi.org/format/#error-objects) work.
 
-### Responses to failed resource creation:
```
Comment (ethanresnick) [inline examples/index.md]: I changed this heading because the errors in this section didn't necessarily have to do with creation (they could also be in response to updates). More importantly, though, I tried to structure the examples from simple -> advanced, so starting with a heading that said "Basic..." made sense

File: examples/index.md [hunk 2]
```diff
@@ -179,63 +181,26 @@ Content-Type: application/vnd.api+json
...
       "title":  "Invalid Attribute",
-      "source": { "pointer": "data/attributes/first-name" },
+      "source": { "pointer": "/data/attributes/first-name" },
       "detail": "First name must contain at least three characters."
     }
   ]
 }
 ```
 
-Notice that the attribute `first-name` is dasherized.  Dasherized attributes are preferred
```
Comment (ethanresnick) [inline examples/index.md]: Removed this paragraph because it's not error specific, and I wanted to keep this section as concise as possible, since it's already so long.

File: examples/index.md [hunk 3]
```diff
@@ -167,9 +167,11 @@ Content-Type: application/vnd.api+json
 
 Examples of how [error objects](http://jsonapi.org/format/#error-objects) work.
 
-### Responses to failed resource creation:
+### A Basic Error Object
```
Comment (ethanresnick) [inline examples/index.md:170]: I changed this heading because the errors in this section didn't necessarily have to do with creation (they could also be in response to updates). More importantly, though, I tried to structure the examples from simple -> advanced, so starting with a heading that said "Basic..." made sense

File: examples/index.md [hunk 4]
```diff
@@ -178,64 +180,35 @@ Content-Type: application/vnd.api+json
...
+      "source": { "pointer": "/data/attributes/first-name" },
       "title":  "Invalid Attribute",
-      "source": { "pointer": "data/attributes/first-name" },
       "detail": "First name must contain at least three characters."
     }
   ]
 }
 ```
 
-Notice that the attribute `first-name` is dasherized.  Dasherized attributes are preferred
```
Comment (ethanresnick) [inline examples/index.md:189]: I removed this note about dasherization because it wasn't specific to errors and I wanted to keep this section (which is already, by necessity, pretty long, as short as possible.)

File: examples/index.md [hunk 5]
```diff
@@ -178,64 +180,35 @@ Content-Type: application/vnd.api+json
...
-over underscored.
-
-Notice also that the `detail` is specific to this occurrence of the problem,
-whereas the `title` is more generic.
-
-Notice that the `source` points to the associated attribute with the error.
```
Comment (ethanresnick) [inline examples/index.md:195]: These notes on `detail`, `title`, and `source` are all introduced again below.

File: examples/index.md [hunk 6]
```diff
@@ -178,64 +180,35 @@ Content-Type: application/vnd.api+json
...
-
-Response with one error object, with relationship error:
-
-```http
-HTTP/1.1 422 Unprocessable Entity
-Content-Type: application/vnd.api+json
-
-{
-  "errors": [
```
Comment (ethanresnick) [inline examples/index.md:204]: This example was essentially identical to the first, except that it included `status`. So I just put `status` in the first example and then removed this one—again, to make this section shorter.

File: examples/index.md [hunk 7]
```diff
@@ -178,64 +180,35 @@ Content-Type: application/vnd.api+json
...
-Content-Type: application/vnd.api+json
+The `status` member represents the HTTP status code associated with the problem.
+It's very helpful when multiple errors are returned at once (see below), as the
+HTTP response itself can only have one status code. However, it can also be
+useful for single errors, to save clients the trouble of consulting the HTTP
+headers, or for using JSON API over non-HTTP protocols, which may be officially
+supported in the near future.
 
-{
-  "errors": [
```
Comment (ethanresnick) [inline examples/index.md:228]: I removed this example because it just doesn't make sense...it's a client side error, while these examples are for errors returned by the server.

File: examples/index.md [hunk 8]
```diff
@@ -246,15 +219,7 @@ Content-Type: application/vnd.api+json
...
-      "detail": "Access to secret powers not authorized on Sundays."
-    },
-    {
-      "status": "404",
-      "links": { "about": {
-        "href": "http://example.com/weapons/vorpal-blade",
-        },
-      },
-      "title": "Weapon not found."
+      "detail": "Editing secret powers is not authorized on Sundays."
```
Comment (ethanresnick) [inline examples/index.md:222]: I changed this `detail` because _accessing_ an attribute can't be a 403, because an attribute doesn't have it's own URI. Instead, you'd have to request a whole resource, and then the 403 would apply to that whole request. Also, the only way I could see these error object realistically be returned together was on a request like `PATCH /wizards/1`, so "editing" made sense for the message.

File: examples/index.md [hunk 9]
```diff
@@ -246,15 +219,7 @@ Content-Type: application/vnd.api+json
     {
       "status": "403",
       "source": { "pointer": "data/attributes/secret-powers" },
-      "detail": "Access to secret powers not authorized on Sundays."
-    },
-    {
-      "status": "404",
```
Comment (ethanresnick) [inline examples/index.md:252]: I removed this example because it's an incorrect use of the `about` link, which is supposed to give a uri for this _particular occurence_ of the problem (so something like `http://example.com/errors/some-uuid`.) Moreover, I'm not sure whether 404 is the right response code (it would depend, I guess, on the request).

I know you wanted a 404 example, so I'd be fine with adding something back in here, we just have to make sure it makes sense.

File: examples/index.md [hunk 10]
```diff
@@ -263,25 +228,17 @@ Content-Type: application/vnd.api+json
     },
     {
       "status": "500",
+      "source": { "pointer": "data/attributes/reputation" },
       "title": "The backend responded with an error",
       "detail": "Reputation service not responding after three requests."
-    },
-    {
-      "status": "400",
-      "detail": "JSON parse error - Expecting property name enclosed in double quotes: line 1 column 2 (char 1)"
```
Comment (ethanresnick) [inline examples/index.md:271]: I removed this error here because there's no way a parse error can co-occur with these other errors in the same response. However, I added this example back under the "Advanced source usage" section.

File: examples/index.md [hunk 11]
```diff
@@ -290,33 +247,40 @@ Content-Type: application/vnd.api+json
...
     {
-      "title": "Invalid Attribute",
       "source": { "pointer": "data/attributes/first-name" },
+      "title": "Invalid Attribute",
       "detail": "First name must contain at least three characters."
     },
     {
-      "title": "Invalid Attribute",
       "source": { "pointer": "data/attributes/first-name" },
+      "title": "Invalid Attribute",
```
Comment (ethanresnick) [inline examples/index.md:256]: This is just me neurotically reordering the properties for consistency. Throughout these examples the order is now: status, code, source, title, detail.

File: examples/index.md [hunk 12]
```diff
@@ -290,33 +247,40 @@ Content-Type: application/vnd.api+json
...
+The `code` member of an error object contains an application-specific code
+representing the type of problem encountered. `code` is similar to `title`
+in that both identify a general type of problem (unlike `detail`, which is
+specific to the particular instance of the problem), but dealing with `code`
+is easier programatically, because the "same" `title` may appear in different
+forms due to localization.
+
+For the example below, imagine the API docs specifed the following mapping:
+
+> | Code | Problem                                                   |
```
Comment (ethanresnick) [inline examples/index.md:278]: Changed these table headings to "Code" and "Problem" because, per the paragraph above, a `code` doesn't necessarily map 1-to-1 to a `title` string (because of localization). Accordingly, I also removed the quotation marks from the values in the second column. Finally, I removed some extraneous codes (again, just to keep the section shorter).

File: examples/index.md [hunk 13]
```diff
@@ -325,54 +289,41 @@ HTTP/1.1 422 Unprocessable Entity
 Content-Type: application/vnd.api+json
 
 {
+  "jsonapi": { "version": "1.0" },
```
Comment (ethanresnick) [inline examples/index.md:292]: Here, I brought the `json-api` member into this example to merge it with the one below, since they were both showing more-or-less the same thing. That is, this top example already showed multiple errors, with distinct codes, on the same field (password).

File: examples/index.md [hunk 14]
```diff
@@ -389,127 +342,61 @@ Content-Type: application/vnd.api+json
...
-```http
-GET /api/posts/1
 ```
-
-```http
-HTTP/1.1 400 Bad Request
-Content-Type: application/vnd.api+json
-
 {
-  "errors": [
```
Comment (ethanresnick) [inline examples/index.md:415]: This query parameter stuff re-introduced below.

File: examples/index.md [hunk 15]
```diff
@@ -389,127 +342,61 @@ Content-Type: application/vnd.api+json
...
-        "detail": "Operation missing required keys: 'path'"
-      }
-    ]
-  }
-]
-```
-
-Notice that JSON PATCH is a special semantics and does not mean an HTTP PATCH request
-containing JSON.
+#### Invalid Query Parameters
```
Comment (ethanresnick) [inline examples/index.md:370]: This is now a subsection of "Advanced source usage".

File: examples/index.md [hunk 16]
```diff
@@ -389,127 +342,61 @@ Content-Type: application/vnd.api+json
...
 Content-Type: application/vnd.api+json
 
-
 {
   "errors": [
     {
-      "status": "404",
-      "source": { "pointer": "/data/articles/1" },
-      "detail": "Article with id 1 could not be found."
+      "source": { "parameter": "include" },
```
Comment (ethanresnick) [inline examples/index.md:386]: I switched this to using an invalid `?include` value because the server cannot, in fact, just ignore such an error:

> If a server is unable to identify a relationship path or does not support inclusion of resources from a path, it MUST respond with 400 Bad Request.

Given that requirement, showing an `include`-related error, which will be more common, I think, than custom parameter errors, seemed like it would be more helpful.

File: examples/index.md [hunk 17]
```diff
@@ -389,127 +342,61 @@ Content-Type: application/vnd.api+json
...
     }
   ]
 }
 ```
 
-Since requests are transactional, as the [Bulk
-extension](http://jsonapi.org/extensions/bulk/#bulk-operations) describes: *"The
-state of the server MUST NOT be changed by a request if any individual operation fails."*
-Thus, even though one of the objects to update was found, one was not, so none get updated.
+In most cases, JSON API requires the server to return an error when it encounters
```
Comment (ethanresnick) [inline examples/index.md:394]: This is a revised version of the note that was here before, just clarifying the difference between API-specific query params (which may often be ignored) and JSON API–defined ones, which usually can not be.

File: examples/index.md [hunk 18]
```diff
@@ -389,127 +342,61 @@ Content-Type: application/vnd.api+json
...
-> ...a request MUST completely succeed or fail (in a single "transaction").
-> ...any request that involves multiple operations MUST only succeed if all operations are performed
-successfully
-
-Examples of invalid parameters may be `?page[prev]=` (parameter present, but has no value),
-`?include=auther` (invalid parameter value, perhaps a typo), `?felds[author]=`,
-(invalid parameter key, perhaps a typo), `?redirect_to=http%3A%2F%2Fwww.owasp.org` (invalid
-parameter, in this case, a phishing attack), etc.
-
-### Using JSON API extensions
```
Comment (ethanresnick) [inline examples/index.md:440]: I removed this stuff about extensions because it's all still experimental and very likely to change. E.g. the discussion in #795 could both obsolete the current JSON PATCH format and expand the bulk extension syntax.

Comment (ethanresnick) [inline examples/index.md:440]: Update: just saw that you suggested putting the extension errors onto the respective extension pages. That could work, perhaps, as an separate PR, assuming they can be integrated unobtrusively (or in their own section).

**Comment by bf4** [comment]
@ethanresnick glad I knew to look for this. I didn't see the notification in my email or in the sidebar!

File: examples/index.md (hunk 11 above)
Comment (bf4) [inline examples/index.md:256]: I gave some thought to the ordering, and decided since I didn't know if there was a preferred order, I wouldn't impose one. That said, I guess it improves readability a bit, so I'll leave this comment here :)

File: examples/index.md (hunk 13 above)
Comment (bf4) [inline examples/index.md:292]: so, I see you wrote `json-api` there even though the spec is `jsonapi`.  This is going to be a common source of error, that I expect clients will likely accepts both... I've doubled checked that it's `jsonapi` myself :)

The json-schema is

``` json
    "jsonapi": {
      "description": "An object describing the server's implementation",
      "type": "object",
      "properties": {
        "version": {
          "type": "string"
        },
        "meta": {
          "$ref": "#/definitions/meta"
        }
      },
      "additionalProperties": false
    },
```

File: examples/index.md [hunk 19]
```diff
@@ -389,127 +342,61 @@ Content-Type: application/vnd.api+json
 {
   "errors": [
     {
-      "title":  "Missing Data Member",
-      "source": { "pointer": "/data" }
+      "source": { "pointer": "/" },
+      "detail":  "Missing `data` Member at document's top level."
```
Comment (bf4) [inline examples/index.md:346]: Curious why you chose to use detail, rather than title, and I could have sworn the example of this I saw was `pointer: /data`, so I guess that was not precise?

Comment (bf4) [inline examples/index.md:346]: ha, below you explain that.  quick trigger finger

File: examples/index.md (hunk 18 above)
Comment (bf4) [inline examples/index.md:440]: Definitely a separate PR since it would require more discussion :)

File: examples/index.md (hunk 9 above)
Comment (bf4) [inline examples/index.md:252]: ok, but I got the structure right.  

Ok.. I've found the schema is unlinked but online and in the repo at https://jsonapi.org/schema and is actually pretty clear on this point

``` json
    "link": {
      "description": "A link **MUST** be represented as either: a string containing the link's URL or a link object.",
      "oneOf": [
        {
          "description": "A string containing the link's URL.",
          "type": "string",
          "format": "uri"
        },
        {
          "type": "object",
          "required": [
            "href"
          ],
          "properties": {
            "href": {
              "description": "A string containing the link's URL.",
              "type": "string",
              "format": "uri"
            },
            "meta": {
              "$ref": "#/definitions/meta"
            }
          }
        }
      ]
    },

    "attributes": {
      "description": "Members of the attributes object (\"attributes\") represent information about the resource object in which it's defined.",
      "type": "object",
      "patternProperties": {
        "^(?!relationships$|links$)\\w[-\\w_]*$": {
          "description": "Attributes may contain any valid JSON value."
        }
      },
      "additionalProperties": false
    },
```

File: examples/index.md (hunk 7 above)
Comment (bf4) [inline examples/index.md:228]: Hmm, I got that from Ember.Data, but maybe I'm misunderstanding it

https://github.com/emberjs/data/blob/d842890357840700f0ee17f152415be0cba16030/packages/ember-data/lib/adapters/errors.js#L82-L85

``` es6
  if (!Ember.isArray(errors)) {
    Ember.deprecate('`InvalidError` expects json-api formatted errors.', false, { id: 'ds.errors.invalid-error-expects-json-api-format', until: '2.0.0' });
    errors = errorsHashToArray(errors);
  }
  AdapterError.call(this, errors, 'The adapter rejected the commit because it was invalid');

export function AdapterError(errors, message = 'Adapter operation failed') {
  EmberError.call(this, message);

  this.errors = errors || [
    {
      title: 'Adapter Error',
      detail: message
    }
  ];
}

export function errorsHashToArray(errors) {
  let out = [];

  if (Ember.isPresent(errors)) {
    Object.keys(errors).forEach((key) => {
      let messages = Ember.makeArray(errors[key]);
      for (let i = 0; i < messages.length; i++) {
        out.push({
          title: 'Invalid Attribute',
          detail: messages[i],
          source: {
            pointer: `/data/attributes/${key}`
          }
        });
      }
    });
  }

  return out;
```

>   A `DS.InvalidError` is used by an adapter to signal the external API
>   was unable to process a request because the content was not
>   semantically correct or meaningful per the API. Usually this means a
>   record failed some form of server side validation. When a promise
>   from an adapter is rejected with a `DS.InvalidError` the record will
>   transition to the `invalid` state and the errors will be set to the
>   `errors` property on the record.
>   For Ember Data to correctly map errors to their corresponding
>   properties on the model, Ember Data expects each error to be
>   a valid json-api error object with a `source/pointer` that matches
>   the property name. For example if you had a Post model that
>   looked like this.

File: examples/index.md (hunk 9 above)
Comment (ethanresnick) [inline examples/index.md:252]: Yes, the structure was correct :) My objection was just on semantic grounds

File: examples/index.md (hunk 13 above)
Comment (ethanresnick) [inline examples/index.md:292]: Yup, this probably will be a common error :) Oh well...

**Comment by bf4** [comment]
I'm going to merge this since I think it's pretty good, and want to keep as much discussion in the json-api repo as possible :)

File: examples/index.md (hunk 18 above)
Comment (ethanresnick) [inline examples/index.md:440]: Sounds good!

File: examples/index.md (hunk 7 above)
Comment (ethanresnick) [inline examples/index.md:228]: This looks like an error that Ember Data throws (on the client side) when an Ember Data Adapter—which is meant to interface with the JSON API server, but isn't part of it—reports to Ember Data in the wrong format that the server responded with an error. Basically, that `Ember.deprecate` call is an Ember specific thing that is unrelated to the JSON API error format.
